File: apps/api/src/demo/systemDesign.ts
import type { SystemDesignReport } from "@placepro/shared";

export const demoSystemDesignReport: SystemDesignReport = {
  id: "demo-system-design",
  topicId: "instagram",
  topicTitle: "Design Instagram",
  design: "",
  scores: {
    scalability: 78,
    architecture: 82,
    databaseDesign: 74,
    caching: 70,
    security: 76,
    overall: 76,
  },
  dimensionFeedback: [
    {
      dimension: "scalability",
      score: 78,
      feedback:
        "You identified horizontal scaling and CDN usage. Expand on feed fan-out strategy and how you handle celebrity accounts with millions of followers.",
      improvements: [
        "Quantify QPS and storage estimates (posts/day, media size)",
        "Compare fan-out on write vs fan-out on read with trade-offs",
      ],
    },
    {
      dimension: "architecture",
      score: 82,
      feedback:
        "Clear separation of API, feed service, and media pipeline. Add async workers for notifications and transcoding.",
      improvements: [
        "Draw data flow for upload → processing → feed invalidation",
        "Mention API gateway and service discovery",
      ],
    },
    {
      dimension: "databaseDesign",
      score: 74,
      feedback:
        "Relational model for users/posts is reasonable. Specify sharding key (user_id) and how comments/likes are stored at scale.",
      improvements: [
        "Add index strategy for feed queries",
        "Discuss SQL vs NoSQL for social graph edges",
      ],
    },
    {
      dimension: "caching",
      score: 70,
      feedback:
        "Redis for hot feeds mentioned briefly. Detail TTL, cache-aside vs write-through, and invalidation on new posts.",
      improvements: [
        "Cache celebrity feeds separately",
        "Use CDN cache headers for static media",
      ],
    },
    {
      dimension: "security",
      score: 76,
      feedback:
        "Auth and HTTPS covered. Add rate limiting, signed URLs for media, and abuse detection for uploads.",
      improvements: [
        "OAuth2/JWT session flow",
        "Content moderation pipeline for uploads",
      ],
    },
  ],
  summary:
    "Solid mid-level system design with good architectural decomposition. Strengthen quantitative estimates, caching invalidation, and deep-dive on feed generation to reach senior-level depth.",
  strengths: [
    "Identified core microservices and async processing",
    "Mentioned CDN and object storage for media",
    "Considered read-heavy traffic patterns",
  ],
  improvements: [
    "Add back-of-envelope capacity planning",
    "Elaborate database sharding and replication",
    "Detail cache layers and consistency model",
  ],
  recommendations: [
    "Practice drawing sequence diagrams for post upload and feed read",
    "Study fan-out on write vs read (Twitter/Instagram case studies)",
    "Review Grokking System Design for estimation templates",
  ],
  architectureHighlights: [
    "API Gateway → Feed Service / Media Service / User Service",
    "S3 + CloudFront for media; Redis for hot feeds",
    "Kafka for async fan-out and analytics events",
  ],
};

File: apps/ai-service/routers/system_design.py
from fastapi import APIRouter
from pydantic import BaseModel, Field
from typing import Optional
from services.llm import get_llm_response

router = APIRouter()

TOPIC_CONTEXT = {
    "instagram": "Photo/video social network: feeds, media CDN, fan-out, stories, likes/comments.",
    "whatsapp": "Real-time messaging: WebSockets, delivery receipts, groups, E2E encryption, presence.",
    "uber": "Ride marketplace: geo-matching, trip state machine, pricing, live location, payments.",
    "youtube": "Video platform: upload/transcode pipeline, adaptive streaming, recommendations, CDN.",
}


class SystemDesignRequest(BaseModel):
    topic_id: str = Field(min_length=3)
    topic_title: str = "System Design"
    design: str = Field(min_length=50)
    scale_hint: Optional[str] = None
    discussion_points: list[str] = Field(default_factory=list)


def _fallback(req: SystemDesignRequest) -> dict:
    text = req.design.lower()
    words = len(req.design.split())
    base = 58 if words < 150 else 68 if words < 350 else 76

    def dim_score(keywords: list[str]) -> float:
        hits = sum(1 for k in keywords if k in text)
        return min(95.0, base + hits * 5)

    scores = {
        "scalability": dim_score(["scale", "shard", "cdn", "kafka", "queue", "million"]),
        "architecture": dim_score(["service", "api", "gateway", "micro", "event"]),
        "database_design": dim_score(["sql", "nosql", "redis", "index", "replica"]),
        "caching": dim_score(["cache", "redis", "cdn", "ttl"]),
        "security": dim_score(["auth", "encrypt", "jwt", "rate", "tls"]),
        "overall": 0.0,
    }
    scores["overall"] = round(
        sum(scores[k] for k in ["scalability", "architecture", "database_design", "caching", "security"]) / 5,
        1,
    )

    dimensions = [
        ("scalability", "Scalability"),
        ("architecture", "Architecture"),
        ("database_design", "Database Design"),
        ("caching", "Caching"),
        ("security", "Security"),
    ]

    return {
        "scores": scores,
        "dimension_feedback": [
            {
                "dimension": d.replace("_design", "Design").replace("database", "database")
                if d == "database_design"
                else d,
                "score": scores[d],
                "feedback": f"{label}: expand with technologies, trade-offs, and estimates.",
                "improvements": [
                    f"Add more depth on {label.lower()}",
                    "Include failure modes and mitigations",
                ],
            }
            for d, label in dimensions
        ],
        "summary": f"Mock evaluation for {req.topic_title}. AI service offline — heuristic scores applied.",
        "strengths": ["Submitted a structured design", "Covers multiple system components"],
        "improvements": [
            "Add quantitative capacity planning",
            "Deepen database and cache strategy",
            "Clarify security and auth flows",
        ],
        "recommendations": [
            "Use a consistent outline: requirements → APIs → data → scale → trade-offs",
            "Practice one topic per week with timed 45-minute sessions",
        ],
        "architecture_highlights": [],
    }


@router.post("/evaluate")
async def evaluate_system_design(req: SystemDesignRequest):
    context = TOPIC_CONTEXT.get(req.topic_id, req.topic_title)
    points = "\n".join(f"- {p}" for p in req.discussion_points[:8])

    prompt = f"""You are a senior staff engineer conducting a system design interview.
Evaluate the candidate's design for: {req.topic_title}
Context: {context}
Scale: {req.scale_hint or "Large-scale global system"}

Key areas interviewers expect:
{points}

CANDIDATE DESIGN:
{req.design[:12000]}

Score each dimension 0-100 and give DETAILED, actionable feedback (3-5 sentences per dimension).

Return JSON only:
{{
  "scores": {{
    "scalability": 0-100,
    "architecture": 0-100,
    "database_design": 0-100,
    "caching": 0-100,
    "security": 0-100,
    "overall": 0-100
  }},
  "dimension_feedback": [
    {{
      "dimension": "scalability",
      "score": 0-100,
      "feedback": "detailed paragraph on horizontal scaling, bottlenecks, fan-out, estimates",
      "improvements": ["specific improvement 1", "specific improvement 2"]
    }},
    {{
      "dimension": "architecture",
      "score": 0-100,
      "feedback": "detailed paragraph on services, boundaries, sync/async, diagrams",
      "improvements": ["..."]
    }},
    {{
      "dimension": "databaseDesign",
      "score": 0-100,
      "feedback": "detailed paragraph on schema, sharding, consistency, indexes",
      "improvements": ["..."]
    }},
    {{
      "dimension": "caching",
      "score": 0-100,
      "feedback": "detailed paragraph on cache layers, TTL, invalidation, CDN",
      "improvements": ["..."]
    }},
    {{
      "dimension": "security",
      "score": 0-100,
      "feedback": "detailed paragraph on auth, encryption, abuse prevention",
      "improvements": ["..."]
    }}
  ],
  "summary": "3-4 sentence executive summary for the candidate",
  "strengths": ["specific strength"],
  "improvements": ["cross-cutting improvement"],
  "recommendations": ["study or practice recommendation"],
  "architecture_highlights": ["notable architectural decision from their answer"]
}}

Be rigorous. overall should reflect weighted average of dimensions. Reference their actual design choices."""

    try:
        result = await get_llm_response(prompt, json_mode=True)
        if result.get("scores"):
            return result
    except Exception:
        pass
    return _fallback(req)
